kamon-scala-future/src/main/scala-2.13/kamon/instrumentation/futures/scala/FutureChainingInstrumentation.scala

```diff
@@ -0,0 +1,141 @@
+package kamon.instrumentation.futures.scala
+
+import kamon.Kamon
+import kamon.context.Context
+import kamon.context.Storage.Scope
+import kamon.instrumentation.context._
+import kamon.instrumentation.futures.scala.CallbackRunnableRunInstrumentation.InternalState
+import kanela.agent.api.instrumentation.InstrumentationBuilder
+import kanela.agent.api.instrumentation.bridge.FieldBridge
+import kanela.agent.libs.net.bytebuddy.asm.Advice
+import kanela.agent.libs.net.bytebuddy.matcher.ElementMatchers.isTypeInitializer
+
+import scala.concurrent.Future
+import scala.util.Try
+
+/**
+  * Ensures that chained transformations on Scala Futures (e.g. future.map(...).flatmap(...)) will propagate the context
+  * set on each transformation to the next transformation.
+  */
+class FutureChainingInstrumentation extends InstrumentationBuilder {
+
+  /**
+    * Captures the current context when a Try instance is created. Since Future's use a Try underneath to handle the
+    * completed value we decided to instrument that instead. As a side effect, all Try instances are instrumented even
+    * if they are not being used in a future, although that is just one extra field that will not be used or visible to
+    * anybody who is not looking for it.
+    */
+  onSubTypesOf("scala.util.Try")
+    .mixin(classOf[HasContext.Mixin])
+    .advise(isConstructor, CaptureCurrentContextOnExit)
+
+  /**
+    * Ensures that if Promise.resolve returns a new Try instance, the captured context will be transferred to that new
+    * instance.
+    */
+  onType("scala.concurrent.impl.Promise")
+    .advise(method("resolve"), CopyContextFromArgumentToResult)
+
+  /**
+    * Captures the scheduling timestamp when a CallbackRunnable is scheduled for execution and then uses the Context
+    * from the completed value as the current Context while the Runnable is executed.
+    */
+  onType("scala.concurrent.impl.Promise$Transformation")
+    .mixin(classOf[HasContext.Mixin])
+    .mixin(classOf[HasTimestamp.Mixin])
+    .bridge(classOf[InternalState])
+    .advise(isConstructor, CaptureCurrentContextOnExit)
+    .advise(method("run"), CallbackRunnableRunInstrumentation)
+    .advise(method("submitWithValue"), CaptureCurrentTimestampOnEnter)
+
+  /**
+    * In Scala 2.13, all Futures are created by calling .map(...) on Future.unit and if happens that while that seed
+    * Future was initialized there was non-empty current Context, that Context will be tied to all Futures which is
+    * obviously wrong. Little tweak ensures that no Context is retained on that seed Future.
+    */
+  onType("scala.concurrent.Future$")
+    .advise(isTypeInitializer, classOf[CleanContextFromSeedFuture])
```

Comment:
ahhh, `isTypeInitializer`. I think my problems with #10 must have been due to it running mid-static-initializer. Thanks, I can rest easy now.